bootloader/grub.go

```diff
@@ -452,19 +452,28 @@ func staticCommandLineForGrubAssetEdition(asset string, edition uint) string {
 
 // grubBootAssetPath contains the paths for assets in the boot chain.
 type grubBootAssetPath struct {
-	shimBinary string
-	grubBinary string
+	shimBinary         string
+	grubBinary         string
+	fallbackBinary     string
+	shimFallbackBinary string
+	grubFallbackBinary string
 }
 
 // grubBootAssetsForArch contains the paths for assets for different
 // architectures in a map
 var grubBootAssetsForArch = map[string]grubBootAssetPath{
 	"amd64": {
-		shimBinary: filepath.Join("EFI/boot/", "bootx64.efi"),
-		grubBinary: filepath.Join("EFI/boot/", "grubx64.efi")},
+		shimBinary:         filepath.Join("EFI/boot/", "bootx64.efi"),
+		grubBinary:         filepath.Join("EFI/boot/", "grubx64.efi"),
+		fallbackBinary:     filepath.Join("EFI/boot/", "fbx64.efi"),
+		shimFallbackBinary: filepath.Join("EFI/ubuntu/", "shimx64.efi"),
+		grubFallbackBinary: filepath.Join("EFI/ubuntu/", "grubx64.efi")},
```

Comment:
My understanding is that the original rationale is that if UC is booted in recovery mode, the fallback binaries are used, and those binaries are in `EFI/ubuntu`.

If I remember correctly, at some point I also thought that the boot chain would try the assets in `EFI/boot` first, and those would trigger booting from the fallback binaries in `EFI/ubuntu`. So the original boot chain was 5 entries long, ending with the assets in `EFI/ubuntu` (and leading to a very long boot process unless an EFI variable is set to jump to the end), but now I can't find where it was that I read that.

In our case though, it's definitely true that the assets in `EFI/ubuntu` are the primary binaries when EFI variables are set, and if no variables exist we "fall back" to the binaries in `EFI/BOOT`. On the other hand, `fbx64.efi` causes the "fallback" binaries to be booted, which are in `EFI/ubuntu`. So I'm really not sure what to think.

Reply:
My understanding is that if no variable is set or the variables are broken, it would boot `EFI\boot`. This would look through all directories to find `bootx64.csv` or `boot.csv`, then register again the variable, and the boot one of them. So `EFI\ubuntu` is not fallback. The fallback finds it and re-register it.

Reply:
So if `EFI\BOOT\fbx4.efi` exists, the medium is not removable. and `EFI\BOOT` is the fallback. Am I understanding this correctly? https://github.com/rhboot/shim/blob/main/README.fallback

Reply:
But if `EFI\BOOT\fbx4.efi` does not exist, then `EFI\BOOT` is not a fallback.

Reply:
Yes, I think I understand it the same way. Something like this:
```
If boot variables exist and point to an EFI asset that exists:
    Boot that asset
Else if `EFI/BOOT/fb${arch}.efi` exists:
    For subdir in `EFI/`:
        If `BOOT${arch}.CSV` exists in subdir: # (in our case, `EFI/ubuntu/BOOT${arch}.CSV`)
            For entry in `BOOT${arch}.CSV`:
                If `${subdir}/${entry}` exists: # (in our case, `EFI/ubuntu/shim${arch}.efi`)
                    Boot `${subdir}/${entry}`
                    We're done!
                Endif
            Endfor
        Endif
    Endfor
Endif
# Failed to find `BOOT${arch}.CSV` with correct EFI assets in the same subdir
If `EFI/BOOT/BOOT${arch}.efi` exists:
    Boot `EFI/BOOT/BOOT${arch}.efi`
    We're done!
Else:
    Fail to boot :(
Endif    
```

So to me it seems like these should be changed to the following:
```diff
2,3c2
< 	shimBinary:         filepath.Join("EFI/boot/", "bootx64.efi"),
< 	grubBinary:         filepath.Join("EFI/boot/", "grubx64.efi"),
---
> 	fallbackBootBinary: filepath.Join("EFI/boot/", "bootx64.efi"),
5,6c4,5
< 	shimFallbackBinary: filepath.Join("EFI/ubuntu/", "shimx64.efi"),
< 	grubFallbackBinary: filepath.Join("EFI/ubuntu/", "grubx64.efi")},
---
> 	shimBinary:         filepath.Join("EFI/ubuntu/", "shimx64.efi"),
> 	grubBinary:         filepath.Join("EFI/ubuntu/", "grubx64.efi")},
8,9c7
< 	shimBinary:         filepath.Join("EFI/boot/", "bootaa64.efi"),
< 	grubBinary:         filepath.Join("EFI/boot/", "grubaa64.efi"),
---
> 	fallbackBootBinary: filepath.Join("EFI/boot/", "bootaa64.efi"),
11,12c9,10
< 	shimFallbackBinary: filepath.Join("EFI/ubuntu/", "shimaa64.efi"),
< 	grubFallbackBinary: filepath.Join("EFI/ubuntu/", "grubaa64.efi")},
---
> 	shimBinary:         filepath.Join("EFI/ubuntu/", "shimaa64.efi"),
> 	grubBinary:         filepath.Join("EFI/ubuntu/", "grubaa64.efi")},
```
How does this sound?

Additionally, I don't actually see where either grub asset gets invoked. It's not in `BOOT${arch}.CSV`, so my best guess for grub to be called would be:
1. `fb${arch}.efi` is invoked
2. Searches for `BOOT${arch}.CSV`, finds it in `EFI/ubuntu`
3. Because it was found in `EFI/ubuntu`, searches for `grub${arch}.efi` in `EFI/ubuntu`, and finds it --- but how would it know that grub is the bootloader to look for??
4. Runs `EFI/ubuntu/grub${arch}.efi`, which uses `BOOT${arch}.CSV` to populate options

Reply:
On later thought, we actually probably do want to leave a `grub${arch}.efi` in `EFI/boot` within the grub bootloader code. If `EFI/boot/fb${arch}.efi` exists, then we use `EFI/ubuntu/{shim,grub}${arch}.efi`, but if not we use `EFI/boot/{boot,grub}${arch}.efi`. This depends on the gadget version basically.

So, I propose calling the boot and grub assets in `EFI/boot/` "default", where the presence of the fallback binary leads to booting the non-"default" assets from `EFI/ubuntu/`:
```diff
2,6c2,6
< 	shimBinary:         filepath.Join("EFI/boot/", "bootx64.efi"),
< 	grubBinary:         filepath.Join("EFI/boot/", "grubx64.efi"),
< 	fallbackBinary:     filepath.Join("EFI/boot/", "fbx64.efi"),
< 	shimFallbackBinary: filepath.Join("EFI/ubuntu/", "shimx64.efi"),
< 	grubFallbackBinary: filepath.Join("EFI/ubuntu/", "grubx64.efi")},
---
> 	defaultBootBinary: filepath.Join("EFI/boot/", "bootx64.efi"),
> 	defaultGrubBinary: filepath.Join("EFI/boot/", "grubx64.efi"),
> 	fallbackBinary:    filepath.Join("EFI/boot/", "fbx64.efi"),
> 	shimBinary:        filepath.Join("EFI/ubuntu/", "shimx64.efi"),
> 	grubBinary:        filepath.Join("EFI/ubuntu/", "grubx64.efi")},
8,12c8,12
< 	shimBinary:         filepath.Join("EFI/boot/", "bootaa64.efi"),
< 	grubBinary:         filepath.Join("EFI/boot/", "grubaa64.efi"),
< 	fallbackBinary:     filepath.Join("EFI/boot/", "fbaa64.efi"),
< 	shimFallbackBinary: filepath.Join("EFI/ubuntu/", "shimaa64.efi"),
< 	grubFallbackBinary: filepath.Join("EFI/ubuntu/", "grubaa64.efi")},
---
> 	defaultBootBinary: filepath.Join("EFI/boot/", "bootaa64.efi"),
> 	defaultGrubBinary: filepath.Join("EFI/boot/", "grubaa64.efi"),
> 	fallbackBinary:    filepath.Join("EFI/boot/", "fbaa64.efi"),
> 	shimBinary:        filepath.Join("EFI/ubuntu/", "shimaa64.efi"),
> 	grubBinary:        filepath.Join("EFI/ubuntu/", "grubaa64.efi")},
```

What do you think of this @valentindavid @alfonsosanchezbeato ?

Reply:
This naming looks fine to me, thanks (I am happy as far as as the binaries in EFI/ubuntu have the old name! :smile:).

Regarding a previous comment, afaiu it is `fb${arch}.efi` the one actually updating the EFI boot variables from csv files.